File: examples/driverlib/single_core/erad/debugger_owned_examples/erad_ex2_stack_overflow/erad_ex2_stack_overflow.js
//#############################################################################
//
// erad_ex2_stack_overflow.js
//
// This JavaScript file is used to configure ERAD registers for use with the
// CCS example "erad_ex2_stack_overflow".
//
// Run this ERAD script in the scripting console using the following command
//
// - require("<proj_workspace_path>\\erad_ex2_stack_overflow\\erad_ex2_stack_overflow.js")
//
// This JavaScript file uses CCS Scripting features. For information, please visit:
// https://software-dl.ti.com/ccs/esd/documents/users_guide_ccs_20.0.0/index_debug-scripting.html
//
//#############################################################################

const ERAD = require("./erad.js");

let { ERAD_EBC_BASE, ERAD_Ebc_Regs_Offset, ERAD_Owner, ERAD_Misc } = ERAD;
let { ENABLE } = ERAD_Misc

//
// Launch a scripting session
//
const { initScripting } = require(require.resolve("scripting", { paths: [process.cwd()] }));
const ds = initScripting();
const debugSession = ds.openSession("C29xx_CPU1");

//
// Variables for data read/write sizes
//
var DATA_SIZE_16 = 16;
var DATA_SIZE_32 = 32;

//*****************************************************************************
// Example: Stack Overflow
//*****************************************************************************

var Ebc = ERAD_EBC_BASE[6]

//
// Set owner as debugger
//
debugSession.memory.write(Ebc + ERAD_Ebc_Regs_Offset.EBC_OWNER, ERAD_Owner.ERAD_OWNER_DEBUGGER, DATA_SIZE_32)

//
// Ensure that EBC is in idle mode
//
var ebcStatus = Number(debugSession.memory.readOne(Ebc + ERAD_Ebc_Regs_Offset.EBC_STATUS, DATA_SIZE_32))

if(0 == (ebcStatus &  0x300))
{
    console.log("\nProfile Setup\n");

    //
    // Reset EBC
    //
    debugSession.memory.write(Ebc + ERAD_Ebc_Regs_Offset.EBC_STATUSCLEAR, 0x1, DATA_SIZE_32);
    debugSession.memory.write(Ebc + ERAD_Ebc_Regs_Offset.EBC_CNTL, 0x0, DATA_SIZE_32);
    debugSession.memory.write(Ebc + ERAD_Ebc_Regs_Offset.EBC_REFL, 0x0, DATA_SIZE_32);
    debugSession.memory.write(Ebc + ERAD_Ebc_Regs_Offset.EBC_REFH, 0x0, DATA_SIZE_32);

    //
    // Grab maximum value for stack
    //
  	var stackEnd = Number(debugSession.expressions.evaluate("__TI_STACK_END"))
    console.log("\nSP END: " + stackEnd.toString(16) + "\n");

    //
    // Set EBC_REF to be stackEnd - threshold (0x10) for EBC
    //
    debugSession.memory.write(Ebc + ERAD_Ebc_Regs_Offset.EBC_REFL, stackEnd - 0x10, DATA_SIZE_32);

    //
    // Clear mask value
    //
    debugSession.memory.write(Ebc + ERAD_Ebc_Regs_Offset.EBC_MASKL, 0x0, DATA_SIZE_32);
    debugSession.memory.write(Ebc + ERAD_Ebc_Regs_Offset.EBC_MASKH, 0x0, DATA_SIZE_32);

    //
    // Set CPU to halt when write to address greater than value in EBC_REF
    // occurs
    // ENABLE         = (1  <<  0) | -> Enable the EBC instance
    // BUS_SEL        = (1  <<  1) | -> DWAB for write data accesses
    // HALT           = (1  <<  6) | -> Halt CPU on match
    // INTERRUPT      = (0  <<  7) | -> Do not generate interrupt
    // NMI_EN         = (0  <<  8) | -> Do not generate NMI
    // COMP_MODE      = (0  <<  9) | -> Generate match when address matches EBC_REF
    // STACK_QUAL     = (1  << 12) | -> Enable Stack access qualifier
    // SPSEL_MATCH_EN = (0  << 13) | -> SPSEL match disabled
    // SPSEL          = (0  << 14)
    // EBC_CNTL = 0x1043
    //
    debugSession.memory.write(Ebc + ERAD_Ebc_Regs_Offset.EBC_CNTL, 0x1043, DATA_SIZE_32);

    //
    // Loop until Stack Overflow occurs
    //
    console.log("\nProfile Started. CPU will halt on stack overflow.\n");

    var coreRan = 0;

    while(1)
    {
        //
        // Read status and check for event fired bit to be set active. This
        // signals that a "match" has occured, as defined by the values set in
        // EBC_CNTL and EBC_REF for EBC. In this case, the event fired bit
        // represents whether or not a write has occurred to the address in
        // EBC_REF, which is set to just before __TI_STACK_END value
        //
        var bp_status = Number(debugSession.memory.readOne(Ebc + ERAD_Ebc_Regs_Offset.EBC_STATUS, DATA_SIZE_32))

        if(bp_status & 0x1 == 1)
        {
            console.log("\nSTACK OVERFLOW detected. Halting CPU.\n");
            break;
        }

        //
        // Halt script if CPU is halted
        //
        var halted = debugSession.target.isHalted();

        if(!halted)
        {
            coreRan = 1;
        }

        if(coreRan && halted)
        {
            break;
        }
    }

    //
    // Disable EBC
    //
    EBC_CNTL_RD = Number(debugSession.memory.readOne((Ebc + ERAD_Ebc_Regs_Offset.EBC_CNTL) & ENABLE, DATA_SIZE_32))
    disableEBC = EBC_CNTL_RD & (~ENABLE);
    debugSession.memory.write(Ebc + ERAD_Ebc_Regs_Offset.EBC_CNTL, disableEBC, DATA_SIZE_32);

    console.log("\nProfile stopped\n");
}
else
{
    console.log("\nFAILED. EBC is not in idle state\n");
}

//
// End of File
//

File: examples/driverlib/single_core/erad/debugger_owned_examples/erad_ex2_stack_overflow/erad_ex2_stack_overflow.c
//#############################################################################
//
// FILE:   erad_ex2_stack_overflow.c
//
// TITLE:  ERAD Stack Overflow Example
//
//! \addtogroup driver_example_list
//! <h1> ERAD Stack Overflow </h1>
//!
//! This example uses BUSCOMP6 to monitor the stack. The Bus comparator is
//! set to monitor the data write access bus and generate an interrupt to
//! CPU when a write is detected to end of the STACK within a threshold.
//! A recursive function call is intentionally made to allow the stack
//! to overflow. The example is built with optimization=O0 and might not
//! function as intended with higher optimizations.
//! The included JavaScript file, erad_ex2_stack_overflow.js, programs 
//! ERAD registers in order to detect the stack overflow and halt the CPU
//! once the pointer reaches the threshold at the end of stack.
//!
//! To run the ERAD script, use the following command in the scripting console:
//! - require("<proj_workspace_path>\\erad_ex2_stack_overflow\\erad_ex2_stack_overflow.js")
//!
//! Note that the script must be run after loading the .out on the C29x core. 
//!
//! This JavaScript file uses CCS Scripting features. For information on
//! using the CCS scripting, please visit:
//! https://software-dl.ti.com/ccs/esd/documents/users_guide_ccs_20.0.0/index_debug-scripting.html
//!
//! This example uses 1 HW watchpoint :
//!  - BUSCOMP6 : Data Write Address Bus = Stack end address - 0x10U
//!
//! \b External \b Connections \n
//!  - None.
//!
//! \b Profiling \b Script \b Output
//!  - "STACK OVERFLOW detected. Halting CPU." will be printed in the scripting
//!    console when a stack overflow occurs (that is, when the watchpoint is hit)
//
//#############################################################################
//
//
// //
//	Copyright: Copyright (C) Texas Instruments Incorporated
//	All rights reserved not granted herein.
//
//  Redistribution and use in source and binary forms, with or without 
//  modification, are permitted provided that the following conditions 
//  are met:
//
//  Redistributions of source code must retain the above copyright 
//  notice, this list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright
//  notice, this list of conditions and the following disclaimer in the 
//  documentation and/or other materials provided with the   
//  distribution.
//
//  Neither the name of Texas Instruments Incorporated nor the names of
//  its contributors may be used to endorse or promote products derived
//  from this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
//  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
//  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
//  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT 
//  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT 
//  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
//  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
//  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT 
//  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//

//#############################################################################

//
// Included Files
//
#include "driverlib.h"
#include "device.h"


//
// Globals
//
extern uint32_t __TI_STACK_END;
volatile uint32_t functionCallCount = 0;

//
// Function Prototypes
//
void recursiveFunction(uint32_t delay);

//
// Main
//
int main(void)
{
    //
    // Initializes device clock and peripherals
    //
    Device_init();

    recursiveFunction(1);

    //
    // IDLE loop. Just sit and loop forever (optional)
    //
    while(1)
    {

    }
}

//
// recursive function to fill the stack
//
void recursiveFunction(uint32_t delay)
{
    functionCallCount++;

    //
    // Recursive function
    //
    recursiveFunction(delay + 1UL);
}


//
// End of File
//